src/Chainweb/Pact/Backend/Compaction.hs

```diff
@@ -0,0 +1,726 @@
+
+  withLevel :: LogLevel -> CompactM x -> CompactM x
+  withLevel l = local (set (ceLogger.setLoggerLevel) l)
+
+  withPolicy :: LogPolicy -> CompactM x -> CompactM x
+  withPolicy p = local (set (ceLogger.setLoggerPolicy) p)
+
+-- | Run compaction monad, see 'mkCompactEnv'.
+runCompactM :: CompactEnv -> CompactM a -> IO a
+runCompactM e a = runReaderT (unCompactM a) e
+
+-- | Prepare/Execute a "$VTABLE$"-templated query.
+execM_ :: ()
+  => Text -- ^ query name (for logging purposes)
+  -> TableName -- ^ table name
+  -> Text -- ^ "$VTABLE$"-templated query
+  -> CompactM ()
+execM_ msg tbl q = do
+  logQueryDebug msg
+  q' <- templateStmt tbl q
+  withDb $ \db -> liftIO $ Pact.exec_ db q'
+
+execNoTemplateM_ :: ()
+  => Text -- ^ query name (for logging purposes)
+  -> Utf8 -- ^ query
+  -> CompactM ()
+execNoTemplateM_ msg qry = do
+  logQueryDebug msg
+  withDb $ \db -> liftIO $ Pact.exec_ db qry
+
+-- | Prepare/Execute a "$VTABLE$"-templated, parameterised query.
+--   The parameters are the results of the 'CompactM' 'SType' computations.
+execM' :: ()
+  => Text -- ^ query name (for logging purposes)
+  -> TableName -- ^ table name
+  -> Text -- ^ "$VTABLE$"-templated query
+  -> [SType] -- ^ parameters
+  -> CompactM ()
+execM' msg tbl stmt ps = do
+  logQueryDebug msg
+  stmt' <- templateStmt tbl stmt
+  withDb $ \db -> liftIO $ Pact.exec' db stmt' ps
+
+exec_ :: ()
+  => Text
+  -> Database
+  -> Utf8
+  -> CompactM ()
+exec_ msg db qry = do
+  logQueryDebug msg
+  liftIO $ Pact.exec_ db qry
+
+qry_ :: ()
+  => Text
+  -> Database
+  -> Utf8
+  -> [RType]
+  -> CompactM [[SType]]
+qry_ msg db qry rs = do
+  logQueryDebug msg
+  liftIO $ Pact.qry_ db qry rs
+
+-- | Prepare/Execute a "$VTABLE$"-templated, parameterised query.
+--   'RType's are the expected results.
+qryM :: ()
+  => Text -- ^ query name (for logging purposes)
+  -> TableName -- ^ table name
+  -> Text -- ^ "$VTABLE$"-templated query
+  -> [SType] -- ^ parameters
+  -> [RType] -- ^ result types
+  -> CompactM [[SType]]
+qryM msg tbl q ins outs = do
+  logQueryDebug msg
+  q' <- templateStmt tbl q
+  withDb $ \db -> liftIO $ Pact.qry db q' ins outs
+
+qryNoTemplateM :: ()
+  => Text -- ^ query name (for logging purposes)
+  -> Utf8 -- ^ query
+  -> [SType] -- ^ parameters
+  -> [RType] -- ^ results
+  -> CompactM [[SType]]
+qryNoTemplateM msg q ins outs = do
+  logQueryDebug msg
+  withDb $ \db -> liftIO $ Pact.qry db q ins outs
+
+logQueryDebug :: Text -> CompactM ()
+logQueryDebug msg = do
+  logg Info ("Query: " <> msg)
+
+-- | Statements are templated with "$VTABLE$" substituted
+-- with the currently-focused versioned table.
+templateStmt :: TableName -> Text -> CompactM Utf8
+templateStmt (TableName (Utf8 tblName)) s
+    | tblTemplate `Text.isInfixOf` s =
+        pure $ Utf8 $ Text.encodeUtf8 $
+          Text.replace tblTemplate ("[" <> Text.decodeUtf8 tblName <> "]") s
+    | otherwise = pure $ Utf8 $ Text.encodeUtf8 s
+  where
+    tblTemplate = "$VTABLE$"
```

Comment:
IMO this is not a template but a template variable, so it should be called that.